src/mpid/common/shm/mpidu_init_shm.c

```diff
@@ -288,11 +290,42 @@ int MPIDU_Init_shm_query(int local_rank, void **target_addr)
     MPIR_Assert(MPIDU_Init_shm_local_size > 1);
 
     MPIR_Assert(local_rank < MPIDU_Init_shm_local_size);
-    *target_addr = (char *) baseaddr + local_rank * sizeof(MPIDU_Init_shm_block_t);
+    *target_addr = (char *) baseaddr + local_rank * MPIDU_INIT_SHM_BLOCK_SIZE;
 
     MPIR_FUNC_EXIT;
 
     return mpi_errno;
 }
 
+int MPIDU_Init_shm_atomic_count(void)
+{
+    return MPL_atomic_load_int(&barrier->alloc_count);
+}
```

Comment:
Same comment from previous PR about naming for this function. it looks like the return value can be 0 or 1, so "count" doesn't quite seem to fit.

Reply:
How about `bool MPIDU_Init_shm_atomic_key_exsit()`?

We only allow 1 atomic key, thus we don't really need a key name. I am hoping that we will only ever need 1 atomic key, but we can extend when future need arise.

Reply:
A more elegant method is via PMI if PMI supports put/get without fence and barrier -- which necessarily forces the `get` to be blocking. Does PMIx supports this? We can extend PMI-1 in `hydra` to support this. Then, we no longer need `Init_shm` for bootstrapping; it will be kept as a fallback.

Reply:
See the last fixup commit.

Reply:
> A more elegant method is via PMI if PMI supports shared-domain put/get without fence and barrier -- which necessarily forces the `get` to be blocking. Does PMIx supports this? We can extend PMI-1 in `hydra` to support this. Then, we no longer need `Init_shm` for bootstrapping; it will be kept as a fallback.

I expect PMIx does support it, but haven't tried.